### ISSUE TITLE
download.file and FTP gateway issues

### ISSUE THREAD

**Issue by serbinsh** [PecanProject/pecan#994]
@robkooper 

I have been running into new issues with FTPs from within BNL that relate to our firewall and the need to use a FTP gateway when connecting to an external FTP such as NARR or fluxnet, but not Ameriflux which goes over https.  The issue is that I need to connect via an FTP gateway or it can't reach the external FTP site:

```
> remotefunc <- function() {
+ PEcAn.data.atmosphere::download.NARR('/data/Model_Output/pecan.output/dbfiles/NARR','2004/01/01','2004/12/31')
+ }
> remoteout <- remotefunc()
Loading required package: PEcAn.utils
Loading required package: abind
Loading required package: plyr
Loading required package: XML
Loading required package: ggplot2
Loading required package: randtoolbox
Loading required package: rngWELL
This is randtoolbox. For overview, type 'help("randtoolbox")'.
Loading required package: RCurl
Loading required package: bitops
Loading required package: lubridate

Attaching package: 'lubridate'

The following object is masked from 'package:plyr':

    here

The following object is masked from 'package:base':

    date

2016-08-18 17:17:59 DEBUG  [PEcAn.data.atmosphere::download.NARR] : 
   File ' /data/Model_Output/pecan.output/dbfiles/NARR/pres.sfc.2004.nc ' 
   already exists, skipping to next file. 
trying URL 'ftp://ftp.cdc.noaa.gov/Datasets/NARR/monolevel/dswrf.2004.nc'
Error in download.file(url, new.file) : 
  cannot open URL 'ftp://ftp.cdc.noaa.gov/Datasets/NARR/monolevel/dswrf.2004.nc'
Calls: remotefunc -> <Anonymous> -> download.file
Execution halted
[1] 1
Error in new.files$file : $ operator is invalid for atomic vectors
Calls: do.conversions ... <Anonymous> -> dbfile.input.insert -> basename -> dirname
> proc.time()
   user  system elapsed 
  4.843   0.297   5.399 

```

I noticed that download.file()  uses curl.  

```
./web/PEcAn.Network.R:    download.file(sync.url,destfile=temporaryFile, method="curl")
```

I wonder if you might have an idea of what to use instead on modex to get around FTP gateways?  Maybe it isn't possible.  I am trying to find out how to connect in one command but also using our FTP gateway, in which case I would just need to setup a special case of the download.file() function that would run for networks like ours....

**Comment by robkooper** [comment]
Not sure but one thing to look at is ftp.proxy (http://www.ftpproxy.org/)

**Comment by serbinsh** [comment]
I will see if I can get this to work for us.

S
------------------------------------------------------------------------------
Shawn P. Serbin
Biological, Environmental & Climate Sciences Department
Brookhaven National Laboratory
Upton, NY 11973-5000

Phone: +1 631-344-3165
Fax: +1 631-344-2060
Email: sserbin@bnl.gov<mailto:sserbin@bnl.gov>
Bio:  http://www.bnl.gov/envsci/bio/serbin-shawn.php
TEST Group: http://www.bnl.gov/TEST/
ORCiD: http://orcid.org/0000-0003-4136-8971
Twitter: @doctorjackpine, @test_bnl
------------------------------------------------------------------------------

On Aug 26, 2016, at 11:28 AM, Rob Kooper <notifications@github.com<mailto:notifications@github.com>> wrote:


Not sure but one thing to look at is ftp.proxy (http://www.ftpproxy.org/)

**Comment by serbinsh** [comment]
@robkooper I am going to give this a try: https://blog.linuxnet.ch/ftp-via-proxy/#NcFTP

**Comment by serbinsh** [comment]
It seems this is the solution:

From out IT guy

```
Hi Shawn,

Ok, I have it working.  Basically, you will need to modify the ~/.ncftp/firewall config file for the user that will be running ncftp and add the following:

1)
firewall-type=1
firewall-host=ftpgateway.sec.bnl.local

ftpgateway.sec.bnl.local is the internal-facing interface for the ftpgatewayas opposed to the external which is ftpgateway.bnl.gov. I'm assuming your ftp connections will always initiate from modex, in which case, leave it set to ftpgateway.sec.bnl.local.

2)
Comment out:
##firewall-user=rjperez
##firewall-password=fwpass

as the ftp gateway (or firewall) does not require a login.

3)
finally, you can modify the firewall-exception-list to include internal ftp sites (if they are used).  For example,
firewall-exception-list=.itd.bnl.gov,.bnl.gov,localhost,localdomain


With this config file set up, you should be able to run:
   $ ncftp user@some.ftp.server

and it should automatically connect via the firewall.

Hope this helps.

- Rafael


```


Then to get files use instead ncftp, such as:

```
[sserbin@modex ~]$ ncftpget ftp://ftp:sserbin@ftp.fluxdata.org/.ameriflux_downloads/BASE-BADM/AMF_US-Ha1_BASE-BADM_10-1.zip
AMF_US-Ha1_BASE-BADM_10-1.zip:                           7.84 MB  785.69 kB/s
```

So @jsimkins2 @ankurdesai @robkooper @mdietze Should we enable sever specific options that can be invoked to use different clients for FTP?  I could envision that based on the machine we could have an alternative FTP get option.....maybe this is just modex but could be a problem for other national labs, etc

**Comment by serbinsh** [comment]
I can mimic download.file with ncftpget using something like

ncftpget -c ftp://ftp:sserbin@ftp.fluxdata.org/.ameriflux_downloads/BASE-BADM/AMF_US-Ha1_BASE-BADM_10-1.zip > /data/sserbin/Data/test2.zip

**Comment by serbinsh** [comment]
system("ncftpget -c ftp://ftp:sserbin@ftp.fluxdata.org/.ameriflux_downloads/BASE-BADM/AMF_US-Ha1_BASE-BADM_10-1.zip > /data/sserbin/Data/test2.zip")

**Comment by serbinsh** [comment]
Fixed in #1296
